File: BQKit/Tool/BQImagePicker.m
//
//  BQImagePicker.m
//  NavgationBar-Test
//
//  Created by baiqiang on 16/10/17.
//  Copyright © 2016年 baiqiang. All rights reserved.
//

#import "BQImagePicker.h"

@interface BQImagePicker()<UIImagePickerControllerDelegate,UINavigationControllerDelegate>
/**  选取头像回调block */
@property (nonatomic, copy) void(^handleBlock)(UIImage *image);
/**  图片选择器 */
@property (nonatomic, strong) UIImagePickerController * picker;
/**  裁剪比例 */
@property (nonatomic, assign) ClipSizeType type;
@end

@implementation BQImagePicker
#pragma mark - Class Method
+  (instancetype)allocWithZone:(struct _NSZone *)zone {
    static BQImagePicker * imagePicker;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        imagePicker = [super allocWithZone:zone];
    });
    return imagePicker;
}
+ (void)showPickerImageWithHandleImage:(void (^)(UIImage *))handle {
    [self showPickerImageWithClipType:ClipSizeTypeNone handleImage:handle];
}
+ (void)showPickerImageWithClipType:(ClipSizeType)type handleImage:(void (^)(UIImage *))handle {
    BQImagePicker * picker = [[self alloc] init];
    picker.handleBlock = handle;
    picker.type = type;
    if ([[UIDevice currentDevice].systemVersion floatValue] >= 8.0) {
        UIAlertController *alertVc = [UIAlertController alertControllerWithTitle:@"选择图像" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
        if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
            [alertVc addAction:[UIAlertAction actionWithTitle:@"拍照" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [picker createImagePickerVcWithType:UIImagePickerControllerSourceTypeCamera];
            }]];
        }
        [alertVc addAction:[UIAlertAction actionWithTitle:@"从相册选择" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [picker createImagePickerVcWithType:UIImagePickerControllerSourceTypePhotoLibrary];
        }]];
        [alertVc addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil]];
        
        [[picker currentViewController] presentViewController:alertVc animated:YES completion:nil];
    }
}
#pragma mark - create Class

#pragma mark - Live Cycle

#pragma mark - instancetype Method
- (void)createImagePickerVcWithType:(UIImagePickerControllerSourceType)type{
    self.picker.sourceType = type;
    self.picker.allowsEditing = self.type == ClipSizeTypeNone;
    [[self currentViewController] presentViewController:self.picker animated:YES completion:nil];
}
- (UIViewController *)currentViewController {
    UIViewController * vc = [UIApplication sharedApplication].keyWindow.rootViewController;
    while (vc.presentedViewController) {
        vc = vc.presentedViewController;
    }
    return vc;
}
#pragma mark - LoadNetWrokData

#pragma mark - button Action

#pragma mark - Delegate Method
-(void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    //获取编辑后的图片
    UIImage *image = info[@"UIImagePickerControllerEditedImage"];
    if (!image) {
        image = info[UIImagePickerControllerOriginalImage];
    }
    if (self.type != ClipSizeTypeNone) {
        __weak typeof(self) weakSelf = self;
        [picker dismissViewControllerAnimated:YES completion:^{
            [weakSelf showClipVcWithImage:image];
        }];
    }else {
        [picker dismissViewControllerAnimated:YES completion:nil];
        if (self.handleBlock != nil) {
            self.handleBlock(image);
        }
    }
}
- (void)showClipVcWithImage:(UIImage *)image {
    __weak typeof(self) weakSelf = self;
    [BQDisImageView showClipViewWithImage:image clipSize:self.type callBack:^(UIImage *image) {
        weakSelf.handleBlock(image);
    }];
}
#pragma mark - View Create

#pragma mark - set Method

#pragma mark - get Method
- (UIImagePickerController *)picker {
    if (_picker == nil) {
        _picker = ({
            UIImagePickerController *picker = [[UIImagePickerController alloc] init];
            picker.delegate = self;
            picker;
        });
    }
    return _picker;
}
@end


@interface BQDisImageView ()
/**  图片视图 */
@property (nonatomic, strong) UIImageView *imageView;
/**  背景视图 */
@property (nonatomic, strong) UIView *backView;
/**  图片 */
@property (nonatomic, strong) UIImage *image;
/**  框位置 */
@property (nonatomic, strong) CAShapeLayer * clipLayer;
/**  回调函数 */
@property (nonatomic, copy) void(^callBack)(UIImage * image);
@end

@implementation BQDisImageView

#pragma mark - 类方法
+ (void)showClipViewWithImage:(UIImage *)image clipSize:(ClipSizeType)type callBack:(void (^)(UIImage *))callBack {
    BQDisImageView * disImage = [[self alloc] initWithImage:image clipSize:type];
    disImage.callBack = callBack;
    [[UIApplication sharedApplication].keyWindow addSubview:disImage];
}
#pragma mark - 创建方法
- (instancetype)initWithImage:(UIImage *)image clipSize:(ClipSizeType)size{
    self = [super init];
    if (self != nil) {
        self.image = image;
        [self initUI];
        self.clipLayer = [CAShapeLayer layer];
        
        CGFloat width = [UIScreen mainScreen].bounds.size.width / 4;
        CGFloat height = 0;
        if (size == ClipSizeTypeOneScaleOne) {
            width = width * 4 / 3;
            height = width;
        }else if (size == ClipSizeTypeTwoScaleOne) {
            height = width;
            width *= 2;
        }else {
            height = width * 2;
            width *= 3;
        }
        
        UIColor * layerColor = [UIColor greenColor];
        self.clipLayer.frame = CGRectMake(0, 0, width, height);
        self.clipLayer.position = self.center;
        self.clipLayer.borderWidth = 1.0f;
        self.clipLayer.borderColor = layerColor.CGColor;
        CGFloat spacing = 20;
        UIBezierPath * path = [UIBezierPath bezierPath];
        
        [path moveToPoint:CGPointMake(0, 20)];
        [path addLineToPoint:CGPointZero];
        [path addLineToPoint:CGPointMake(spacing, 0)];
        
        [path moveToPoint:CGPointMake(width - spacing, 0)];
        [path addLineToPoint:CGPointMake(width, 0)];
        [path addLineToPoint:CGPointMake(width, spacing)];
        
        [path moveToPoint:CGPointMake(width, height - spacing)];
        [path addLineToPoint:CGPointMake(width, height)];
        [path addLineToPoint:CGPointMake(width - spacing, height)];
        
        [path moveToPoint:CGPointMake(20, height)];
        [path addLineToPoint:CGPointMake(0, height)];
        [path addLineToPoint:CGPointMake(0, height - 20)];
        
        self.clipLayer.path = path.CGPath;
        self.clipLayer.lineWidth = 3.0f;
        self.clipLayer.strokeColor = layerColor.CGColor;
        self.clipLayer.fillColor = [UIColor clearColor].CGColor;
        
        [self.layer addSublayer:self.clipLayer];
    }
    return self;
}

#pragma mark - 实例方法

- (void)initUI{
    
    self.backgroundColor = [UIColor grayColor];
    
    CGFloat width = [UIScreen mainScreen].bounds.size.width;
    CGFloat height = [UIScreen mainScreen].bounds.size.height;
    self.frame = CGRectMake(0, 0, width, height);
    UIImageView *imageView = [[UIImageView alloc] initWithImage:self.image];
    imageView.userInteractionEnabled = YES;
    imageView.frame = CGRectMake(0, 0, width, imageView.image.size.height * width / imageView.image.size.width);
    imageView.center = self.center;
    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(gestureRecognizerChange:)];
    [imageView addGestureRecognizer:pan];
    UIPinchGestureRecognizer *pinch = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(gestureRecognizerChange:)];
    [imageView addGestureRecognizer:pinch];
    self.imageView = imageView;
    [self addSubview:imageView];
    
    [self createBtnWithFrame:CGRectMake(40, height - 70, 50, 30) title:@"取消" tag:100];
    [self createBtnWithFrame:CGRectMake(width - 90, height - 70, 50, 30) title:@"裁剪" tag:101];
}
- (void)createBtnWithFrame:(CGRect)frame title:(NSString *)title tag:(NSInteger)tag {
    UIButton  * backBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [backBtn addTarget:self action:@selector(bottomBtnAction:) forControlEvents:UIControlEventTouchUpInside];
    backBtn.frame = frame;
    backBtn.backgroundColor = self.backgroundColor;
    backBtn.tag = tag;
    [backBtn setTitle:title forState:UIControlStateNormal];
    [self addSubview:backBtn];
}
#pragma mark - 事件响应方法
- (void)gestureRecognizerChange:(UIGestureRecognizer*)gesture{
    // 拖拽
    if ([gesture isKindOfClass:[UIPanGestureRecognizer class]]) {
        UIPanGestureRecognizer *pan = (UIPanGestureRecognizer *)gesture;
        static CGPoint startCenter;
        if (pan.state == UIGestureRecognizerStateBegan) {
            startCenter = self.imageView.center;
        }
        else if (pan.state == UIGestureRecognizerStateChanged) {
            // 此处必须从self.view中获取translation，因为translation和view的transform属性挂钩，若transform改变了则translation也会变
            CGPoint translation = [pan translationInView:self];
            self.imageView.center = CGPointMake(startCenter.x + translation.x, startCenter.y + translation.y);
        }
        else if (pan.state == UIGestureRecognizerStateEnded) {
            startCenter = CGPointZero;
            if (self.imageView.frame.origin.x > self.clipLayer.frame.origin.x) {
                [UIView animateWithDuration:0.1 animations:^{
                    CGRect frame = self.imageView.frame;
                    frame.origin.x = self.clipLayer.frame.origin.x;
                    self.imageView.frame = frame;
                }];
            }
            if (CGRectGetMaxX(self.imageView.frame) < CGRectGetMaxX(self.clipLayer.frame)) {
                [UIView animateWithDuration:0.1 animations:^{
                    CGRect frame = self.imageView.frame;
                    frame.origin.x = CGRectGetMaxX(self.clipLayer.frame) - frame.size.width;
                    self.imageView.frame = frame;
                }];
            }
            if (self.imageView.frame.origin.y > self.clipLayer.frame.origin.y) {
                [UIView animateWithDuration:0.1 animations:^{
                    CGRect frame = self.imageView.frame;
                    frame.origin.y = self.clipLayer.frame.origin.y;
                    self.imageView.frame = frame;
                }];
            }
            if (CGRectGetMaxY(self.imageView.frame) < CGRectGetMaxY(self.clipLayer.frame)) {
                [UIView animateWithDuration:0.1 animations:^{
                    CGRect frame = self.imageView.frame;
                    frame.origin.y = CGRectGetMaxY(self.clipLayer.frame) - frame.size.height;
                    self.imageView.frame = frame;
                }];
            }
        }
    }
    // 缩放
    else {
        UIPinchGestureRecognizer *pinch = (UIPinchGestureRecognizer *)gesture;
        static CGFloat startScale;
        if (pinch.state == UIGestureRecognizerStateBegan) {
            startScale = pinch.scale;
        }
        else if (pinch.state == UIGestureRecognizerStateChanged) {
            CGFloat scale = (pinch.scale - startScale) +1;
            self.imageView.transform = CGAffineTransformScale(self.imageView.transform, scale, scale);
            startScale = pinch.scale;
        }
        else if (pinch.state == UIGestureRecognizerStateEnded) {
            startScale = 1;
            if (self.imageView.frame.size.width < self.clipLayer.bounds.size.width || self.imageView.frame.size.height < self.clipLayer.bounds.size.height) {
                
                [UIView animateWithDuration:0.1 animations:^{
                    self.imageView.frame = CGRectMake(0, 0, self.bounds.size.width, self.imageView.image.size.height * self.bounds.size.width / self.imageView.image.size.width);
                    self.imageView.center = self.center;
                }];
                
            }
        }
    }
}
- (void)bottomBtnAction:(UIButton *)btn {
    if (btn.tag == 101) {
        self.clipLayer.hidden = YES;
        CGFloat scale = [UIScreen mainScreen].scale;
        UIGraphicsBeginImageContextWithOptions(self.bounds.size,NO,scale);
        [self.layer renderInContext:UIGraphicsGetCurrentContext()];
        UIImage * img = UIGraphicsGetImageFromCurrentImageContext();
        UIImage * image = [UIImage imageWithCGImage:CGImageCreateWithImageInRect(img.CGImage, CGRectMake(self.clipLayer.frame.origin.x * scale, self.clipLayer.frame.origin.y * scale, self.clipLayer.frame.size.width * scale, self.clipLayer.frame.size.height * scale))];
        UIGraphicsEndImageContext();
        if (self.callBack != nil) {
            self.callBack(image);
        }
    }
    [self removeFromSuperview];
}
#pragma mark - Method

#pragma mark - set方法

#pragma mark - get方法

@end

File: SwiftKit/tool/BQImagePicker.swift
//
//  BQImagePicker.swift
//  HaoJiLai
//
//  Created by baiqiang on 16/11/7.
//  Copyright © 2016年 baiqiang. All rights reserved.
//

import UIKit

enum ClipSizeType {
    case none
    case oneScaleOne
    case twoScaleOne
    case threeScaleTwo
}
/**
 需要在info.plist 文件中添加字段
 Privacy - Photo Library Usage Description //是否允许访问相册
 Privacy - Camera Usage Description        //是否允许是使用相机
 */
class BQImagePicker: NSObject,UIImagePickerControllerDelegate,UINavigationControllerDelegate {
    //MARK: - ***** Ivars *****
    //单例写法
    private static let sharedInstance = BQImagePicker()
    private override init() {
        //        self.imagePicker = UIImagePickerController()
        //        self.imagePicker.delegate = self;
    }
    lazy var imagePicker:UIImagePickerController = {
        let pickVc = UIImagePickerController()
        pickVc.delegate = self;
        return pickVc;
    }()
    var handle:((UIImage) -> Void)?
    var type:ClipSizeType!
    
    //MARK: - ***** Class Method *****
    class func showPicker(handle:((UIImage) -> Void)?) -> Void {
        self.showPicker(type: ClipSizeType.none, handle: handle)
    }
    class func showPicker(type:ClipSizeType, handle:((UIImage) -> Void)?) -> Void {
        let picker = BQImagePicker.sharedInstance
        picker.type = type
        picker.handle = handle
        var strDatas:[String] = []
        if UIImagePickerController.isSourceTypeAvailable(.camera) {
            strDatas.append("拍照")
        }
        strDatas.append("相册")
        BQSheetView.showSheetView(tableDatas: strDatas, title: "获取图像方式") { (index) in
            if strDatas[index] == "拍照" {
                picker.showImagePickVc(type: .camera)
            }else {
                picker.showImagePickVc(type: .photoLibrary)
            }
        }
    }
    //MARK: - ***** initialize Method *****
    private func showImagePickVc(type:UIImagePickerControllerSourceType) -> Void {
        self.imagePicker.sourceType = type
        self.imagePicker.allowsEditing = self.type == ClipSizeType.none
        BQTool.currentVc().present(self.imagePicker, animated: true, completion: nil)
    }
    //MARK: - ***** Instance Method *****
    
    //MARK: - ***** LoadData Method *****
    
    //MARK: - ***** create Method *****
    
    //MARK: - ***** respond event Method *****
    
    //MARK: - ***** Protocol *****
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        let image: UIImage!
        if self.type == ClipSizeType.none {
            image = info["UIImagePickerControllerEditedImage"] as! UIImage
            picker.dismiss(animated: true, completion: {
                if self.handle != nil {
                    self.handle!(image)
                }
            })
        }else {
            image = info["UIImagePickerControllerOriginalImage"] as! UIImage;
            picker.dismiss(animated: true, completion: {
                BQClipView.showClipView(image: image, type: self.type, handle: self.handle)
            })
        }
        
    }
}

class BQClipView: UIView {
    //MARK: - ***** Ivars *****
    var image: UIImage!
    var type: ClipSizeType
    var imageView: UIImageView!
    var clipLayer: CAShapeLayer!
    var handle:((UIImage) -> Void)?
    var startCenter: CGPoint = CGPoint(x: 0, y: 0)
    var startSacle: CGFloat = 0
    //MARK: - ***** Lifecycle *****
    
    //MARK: - ***** Class Method *****
    class func showClipView(image:UIImage, type:ClipSizeType ,handle:((UIImage) -> Void)?) {
        let clipView = BQClipView.init(image: image, type: type)
        clipView.handle = handle
        UIApplication.shared.keyWindow?.addSubview(clipView)
    }
    //MARK: - ***** initialize Method *****
    init(image:UIImage, type:ClipSizeType) {
        //调用父类的构造函数
        self.image = image
        self.type = type
        super.init(frame: UIScreen.main.bounds)
        self.initUI()
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    //MARK: - ***** Instance Method *****
    func initUI() {
        self.backgroundColor = UIColor.gray
        self.imageView = UIImageView(image: self.image)
        self.imageView.frame = CGRect(x: 0, y: 0, width: self.width, height: self.image.size.height * self.width / self.image.size.width)
        self.imageView.center = self.center
        
        self.imageView.isUserInteractionEnabled = true
        let pan = UIPanGestureRecognizer(target: self, action: #selector(gestureRecognizerChange(gesture:)))
        self.imageView.addGestureRecognizer(pan)
        let pinch = UIPinchGestureRecognizer(target: self, action: #selector(gestureRecognizerChange(gesture:)))
        self.imageView.addGestureRecognizer(pinch)
        self.addSubview(self.imageView)
        self.createLayer()
        self.createBtn(frame: CGRect(x: 40, y: self.height - 70, width: 50, height: 30), title: "取消", tag: 100)
        self.createBtn(frame: CGRect(x: self.width - 90, y: self.height - 70, width: 50, height: 30), title: "裁剪", tag: 101)
    }
    
    func createLayer() {
        self.clipLayer = CAShapeLayer()
        var width = self.width / 4.0
        var height: CGFloat = 0
        switch self.type {
        case ClipSizeType.oneScaleOne:
            width = self.width / 3.0
            height = width
            break
        case ClipSizeType.twoScaleOne:
            height = width;
            width *= 2;
            break
        case ClipSizeType.threeScaleTwo:
            height = width * 2;
            width *= 3;
            break
        default:
            break
        }
        
        self.clipLayer.frame = CGRect(x: 0, y: 0, width: width, height: height)
        self.clipLayer.position = self.center
        let length:CGFloat = 20
        let path = UIBezierPath()
        //左上角
        path.move(to: CGPoint(x: 0, y: length))
        path.addLine(to: CGPoint(x: 0, y: 0))
        path.addLine(to: CGPoint(x: length, y: 0))
        //右上角
        path.move(to: CGPoint(x: width - length, y: 0))
        path.addLine(to: CGPoint(x: width, y: 0))
        path.addLine(to: CGPoint(x: width, y: length))
        //右下角
        path.move(to: CGPoint(x: width, y: height - length))
        path.addLine(to: CGPoint(x: width, y: height))
        path.addLine(to: CGPoint(x: width - 20, y: height))
        //左下角
        path.move(to: CGPoint(x: length, y: height))
        path.addLine(to: CGPoint(x: 0, y: height))
        path.addLine(to: CGPoint(x: 0, y: height - length))
        
        self.clipLayer.path = path.cgPath
        self.clipLayer.lineWidth = 3.0
        self.clipLayer.strokeColor = UIColor.green.cgColor
        self.clipLayer.fillColor = UIColor.clear.cgColor
        
        width = self.width
        height = self.height
        self.addMaskLayer(frame: CGRect(x: 0, y: 0, width: width, height: self.clipLayer.frame.origin.y))
        self.addMaskLayer(frame: CGRect(x: 0, y: self.clipLayer.frame.origin.y, width: self.clipLayer.frame.origin.x, height: self.clipLayer.frame.size.height))
        self.addMaskLayer(frame: CGRect(x: 0, y: self.clipLayer.frame.maxY, width: width, height: height - self.clipLayer.frame.maxY))
        self.addMaskLayer(frame: CGRect(x: self.clipLayer.frame.maxX, y: self.clipLayer.frame.origin.y, width: width - self.clipLayer.frame.maxX, height: self.clipLayer.frame.size.height))
        self.layer.addSublayer(self.clipLayer)
    }
    func addMaskLayer(frame: CGRect) {
        let maskLayer = CAShapeLayer()
        maskLayer.frame = frame
        maskLayer.backgroundColor = UIColor(white: 0.2, alpha: 0.7).cgColor
        self.layer.addSublayer(maskLayer)
    }
    //MARK: - ***** LoadData Method *****
    
    //MARK: - ***** create Method *****
    func  createBtn(frame:CGRect, title:String?, tag:Int) {
        let btn = UIButton(type: .custom)
        btn.addTarget(self, action: #selector(btnAction(btn:)), for: UIControlEvents.touchUpInside)
        btn.frame = frame
        btn.tag = tag
        btn.setTitle(title, for: UIControlState.normal)
        self.addSubview(btn)
    }
    //MARK: - ***** respond event Method *****
    func gestureRecognizerChange(gesture:UIGestureRecognizer){
        if gesture.isKind(of: UIPanGestureRecognizer.self) {
            let pan = gesture as! UIPanGestureRecognizer
            switch pan.state {
            case .began:
                self.startCenter = self.imageView.center
                break
            case .changed:
                let translation = pan.translation(in: self)
                self.imageView.center = CGPoint(x: self.startCenter.x + translation.x, y: self.startCenter.y + translation.y)
                break
            case .ended:
                self.startCenter = CGPoint(x: 0, y: 0)
                if self.imageView.frame.origin.x > self.clipLayer.frame.origin.x {
                    UIView.animate(withDuration: 0.1, animations: {
                        var frame = self.imageView.frame;
                        frame.origin.x = self.clipLayer.frame.origin.x;
                        self.imageView.frame = frame;
                    })
                }
                if (self.imageView.frame.maxX < self.clipLayer.frame.maxX) {
                    UIView.animate(withDuration: 0.1, animations: {
                        var frame = self.imageView.frame;
                        frame.origin.x = self.clipLayer.frame.maxX - frame.size.width;
                        self.imageView.frame = frame;
                    })
                }
                if self.imageView.frame.origin.y > self.clipLayer.frame.origin.y {
                    UIView.animate(withDuration: 0.1, animations: {
                        var frame = self.imageView.frame;
                        frame.origin.y = self.clipLayer.frame.origin.y;
                        self.imageView.frame = frame;
                    })
                }
                if (self.imageView.frame.maxY < self.clipLayer.frame.maxY) {
                    UIView.animate(withDuration: 0.1, animations: {
                        var frame = self.imageView.frame;
                        frame.origin.y = self.clipLayer.frame.maxY - frame.size.height;
                        self.imageView.frame = frame;
                    })
                }
                break
            default:
                break
            }
        }else {
            let pinch = gesture as! UIPinchGestureRecognizer
            switch pinch.state {
            case .began:
                startSacle = pinch.scale
                break
            case .changed:
                let scale = (pinch.scale - startSacle) + 1
                
                self.imageView.transform = self.imageView.transform.scaledBy(x: scale, y: scale)
                startSacle = pinch.scale
                break
            case .ended:
                startSacle = 1
                if self.imageView.frame.size.width < self.clipLayer.bounds.size.width || self.imageView.frame.size.height < self.clipLayer.bounds.size.height {
                    UIView.animate(withDuration: 0.1, animations: {
                        self.imageView.frame = CGRect(x: 0,y: 0,width: self.bounds.size.width, height: self.image.size.height * self.width / self.image.size.width);
                        self.imageView.center = self.center;
                    })
                }
                break
            default:
                break
            }
        }
    }
    func btnAction(btn:UIButton) {
        if btn.tag == 101 {
            self.clipLayer.isHidden = true
            let scale = UIScreen.main.scale
            UIGraphicsBeginImageContextWithOptions(self.bounds.size, false, scale)
            self.layer.render(in: UIGraphicsGetCurrentContext()!)
            let img = UIGraphicsGetImageFromCurrentImageContext()!
            let backImage = UIImage(cgImage: img.cgImage!.cropping(to: CGRect(x: self.clipLayer.frame.origin.x * scale, y: self.clipLayer.frame.origin.y * scale, width: self.clipLayer.frame.size.width * scale, height: self.clipLayer.frame.size.height * scale))!)
            UIGraphicsEndImageContext()
            if self.handle != nil {
                self.handle!(backImage)
            }
        }
        self.removeFromSuperview()
    }
}
